### PR TITLE
Allow pay request and hop hints as input to queryroutes

### PR DESCRIPTION
**Pull request by Bluetegu**
This addresses issue #2056 

Two additional optional input flags were added to queryroutes cli command:
payment request (invoice)
hints (in json format)

This is useful mostly to compute routes to private destinations.

The current implementation assumes that the hints provide the last few steps towards the destination, i.e. that their final hop connects to the last node on the path. In particular, hints to cross a private edge in the middle of the network are not supported. This limitation can be lifted in future.

For testing a setup of A->B->C where B->C is private was used.

To test the use of an invoice as input to queryroutes, an invoice was generated on C. The invoice was paid using sendtoroute, with queryroutes feeding it with routes. The invoice was used as input to queryroutes. Here is a log of that command:

ubuntu@btc:~/go/dev/alice$ lncli-a queryroutes --pay_req="lnsb53220n1pd746evpp58njslvvsjfrwljr49aj4qtq8xcnghvppmv03jgmledgmlvlegdvqdzz2pexjanpw3jjq6twwehkjcm9yp6x7grzv5s8qctfvssxy7fqwdjkuer5daex7at5v5cqzysxqyjw5qrzjqf4996kagt4tku8fkpyp4acqr55e7ltmgypsjaamx7304trmhcqpyqq8gvqqqqgqqqqqqfcsqqqqqqqqjql7ga8cm06x68w5uw3glyy37xmqzpncqajprrnan8wes0fx0q8mf4s6x58dqys4fee6xmarm72f3ulhqdg6rg3aumxyg9x8pu78zwwccqnyvlyz" | lncli-a sendtoroute "3ce50fb1909246efc8752f65502c0736268bb021db1f19237fcb51bfb3f94358" -
{
        "payment_error": "",
        "payment_preimage": "94192fedd60bd0e31db8160aeaafac48cf79de5196af77e3bc58a6d2bde9734d",
        "payment_route": {
                "total_time_lock": 2016,
                "total_fees": 10,
                "total_amt": 5332,
                "hops": [
                        {
                                "chan_id": 1540415790579712,
                                "chan_capacity": 1000000,
                                "amt_to_forward": 5322,
                                "fee": 10,
                                "expiry": 1872,
                                "amt_to_forward_msat": 5322000,
                                "fee_msat": 10000,
                                "pub_key": "026a52eadd42eabb70e9b0481af7001d299f7d7b41030977bb37a2faac7bbe0012"
                        },
                        {
                                "chan_id": 2043992116101120,
                                "chan_capacity": 5322,
                                "amt_to_forward": 5322,
                                "expiry": 1872,
                                "amt_to_forward_msat": 5322000,
                                "pub_key": "034a15c028970d4070c60f532018cfdb943b584ba267c83fe333ef1ad70b10feaf"
                        }
                ],
                "total_fees_msat": 10000,
                "total_amt_msat": 5332000
        }
}

To test input of hints to queryroutes, a hint that includes the last B->C hop was used as input. This hint is identical to the hint generated when creating the invoice on C.

ubuntu@btc:~/go/dev/alice$ lncli-a queryroutes --hints='{"route_hints": [
        {
            "hop_hints": [
                {
                    "node_id": "026a52eadd42eabb70e9b0481af7001d299f7d7b41030977bb37a2faac7bbe0012",
                    "chan_id": "2043992116101120",
                    "fee_base_msat": 10000,
                    "fee_proportional_millionths": 0,
                    "cltv_expiry_delta": 144
                }
            ]
        }
    ]}' "034a15c028970d4070c60f532018cfdb943b584ba267c83fe333ef1ad70b10feaf" 333
{
    "routes": [
        {
            "total_time_lock": 2016,
            "total_fees": "10",
            "total_amt": "343",
            "hops": [
                {
                    "chan_id": "1540415790579712",
                    "chan_capacity": "1000000",
                    "amt_to_forward": "333",
                    "fee": "10",
                    "expiry": 1872,
                    "amt_to_forward_msat": "333000",
                    "fee_msat": "10000",
                    "pub_key": "026a52eadd42eabb70e9b0481af7001d299f7d7b41030977bb37a2faac7bbe0012"
                },
                {
                    "chan_id": "2043992116101120",
                    "chan_capacity": "333",
                    "amt_to_forward": "333",
                    "fee": "0",
                    "expiry": 1872,
                    "amt_to_forward_msat": "333000",
                    "fee_msat": "0",
                    "pub_key": "034a15c028970d4070c60f532018cfdb943b584ba267c83fe333ef1ad70b10feaf"
                }
            ],
            "total_fees_msat": "10000",
            "total_amt_msat": "343000"
        }
    ]
}

### CODE REVIEW

File: lnrpc/rpc.proto [hunk 1]
```diff
@@ -1375,6 +1388,13 @@ message Hop {
     string pub_key = 8 [json_name = "pub_key"];
 }
 
+/**
+A set of hops, ordered, as provided in the routing hints of an invoice
+*/
+message Hops {
+    repeated Hop hops = 4 [json_name = "hops"];
```
Comment (halseth) [inline lnrpc/rpc.proto]: why `4`?

File: routing/pathfind.go [hunk 2]
```diff
@@ -874,7 +875,7 @@ func findPaths(tx *bolt.Tx, graph *channeldb.ChannelGraph,
 			// root path removed, we'll attempt to find another
 			// shortest path from the spur node to the destination.
 			spurPath, err := findPath(
-				tx, graph, nil, spurNode, target,
+				tx, graph, additionalEdges, spurNode, target,
```
Comment (halseth) [inline routing/pathfind.go]: needs a rebase to work with new method signature! 😀

File: rpcserver.go [hunk 3]
```diff
@@ -3420,25 +3420,85 @@ func (r *rpcServer) GetNodeInfo(ctx context.Context,
...
-	// we can properly manipulate.
-	pubKeyBytes, err := hex.DecodeString(in.PubKey)
-	if err != nil {
-		return nil, err
+	// If a payment request was specified, extract the destination, amount,
+	// and hops from it.
+	payReq, err := zpay32.Decode(
+		in.PaymentRequest, activeNetParams.Params,
+	)
+	if err != nil && len(in.PaymentRequest) > 0 {
```
Comment (halseth) [inline rpcserver.go]: should check if `PaymentRequest` is present _before_ attempting to parse it.

File: rpcserver.go [hunk 4]
```diff
@@ -3420,25 +3420,85 @@ func (r *rpcServer) GetNodeInfo(ctx context.Context,
...
+					FeeBaseMSat:               h.FeeBaseMsat,
+					FeeProportionalMillionths: h.FeeProportionalMillionths,
+					CLTVExpiryDelta:           uint16(h.CltvExpiryDelta),
+				})
+			}
+			routeHints = append(routeHints, route)
+		}
+	}
+
+	// Parse the hex-encoded public key into a full public key object
```
Comment (halseth) [inline rpcserver.go]: you can do this check before parsing the hop hints?

File: rpcserver.go [hunk 5]
```diff
@@ -3420,25 +3420,85 @@ func (r *rpcServer) GetNodeInfo(ctx context.Context,
...
+	)
+
+	// Route hints can be read either from the invoice or from an
+	// explicit set of route hints provided.
+	if payReq != nil {
+		routeHints = payReq.RouteHints
+	}
+	// Explicit route hints take precedence (overrides) over hints
+	// provided in invoice
+	if len(in.Hints) > 0 {
```
Comment (halseth) [inline rpcserver.go]: this `if` is not necessary

File: rpcserver.go [hunk 6]
```diff
@@ -3420,25 +3420,85 @@ func (r *rpcServer) GetNodeInfo(ctx context.Context,
...
+		for _, rh := range in.Hints {
+			var route []routing.HopHint
+			for _, h := range rh.HopHints {
+				var nodeID *btcec.PublicKey
+				if len(h.NodeId) > 0 {
```
Comment (halseth) [inline rpcserver.go]: if this is not set, shouldn't that be an error?

File: lnrpc/rpc.proto (hunk 1 above)
Comment (Bluetegu) [reply lnrpc/rpc.proto]: Lol. Fixed

File: rpcserver.go (hunk 3 above)
Comment (Bluetegu) [reply rpcserver.go]: Yea. I was taking advantage of Decode handling correctly null inputs. Anyhow, I agree that this is convoluted. Rearranged code to make it more readable.

File: rpcserver.go (hunk 4 above)
Comment (Bluetegu) [reply rpcserver.go]: Yes. Moved up (need parsing of payment first though)

File: rpcserver.go (hunk 5 above)
Comment (Bluetegu) [reply rpcserver.go]: Removed.

File: rpcserver.go (hunk 6 above)
Comment (Bluetegu) [reply rpcserver.go]: Yes. Fixed. Now returning custom errors, to ensure the source of the error is known (i.e. that its from malformed hints).

**Comment by halseth** [comment]
Needs a rebase once #3054 is in.

**Comment by Roasbeef** [comment]
Superseded by #3911.

**Pull request closed without merging**